### ISSUE TITLE
Cypress rewrites window.top.location during test and causes issues

### ISSUE THREAD

**Issue by hiezust** [cypress-io/cypress#16905]
### Current behavior
While running our test, we encounter the following cypress error. We have searched the issues for a solution and we have encountered [this one](https://github.com/cypress-io/cypress/issues/8507). We made use of cypress version 7.3.0 as seen in the stack trace but it seems that we still encounter this issue.

```
Error: Webpack Compilation Error
multi .
Module not found: Error: Can't resolve 'C:\Users\User\Desktop\automation-myProjectName' in 'C:\Users\User\Desktop\automation-myProjectName'
Looked for and couldn't find the file at the following paths:
[C:\Users\User\Desktop\automation-myProjectName]
[C:\Users\User\Desktop\automation-myProjectName.js]
[C:\Users\User\Desktop\automation-myProjectName.json]
[C:\Users\User\Desktop\automation-myProjectName.jsx]
[C:\Users\User\Desktop\automation-myProjectName.mjs]
[C:\Users\User\Desktop\automation-myProjectName.coffee]
[C:\Users\User\Desktop\automation-myProjectName.ts]
[C:\Users\User\Desktop\automation-myProjectName.tsx]
[C:\Users\User\Desktop\automation-myProjectName\index.js]
[C:\Users\User\Desktop\automation-myProjectName\index.js.js]
[C:\Users\User\Desktop\automation-myProjectName\index.js.json]
[C:\Users\User\Desktop\automation-myProjectName\index.js.jsx]
[C:\Users\User\Desktop\automation-myProjectName\index.js.mjs]
[C:\Users\User\Desktop\automation-myProjectName\index.js.coffee]
[C:\Users\User\Desktop\automation-myProjectName\index.js.ts]
[C:\Users\User\Desktop\automation-myProjectName\index.js.tsx]
[C:\Users\User\Desktop\automation-myProjectName\index]
[C:\Users\User\Desktop\automation-myProjectName\index.json]
[C:\Users\User\Desktop\automation-myProjectName\index.jsx]
[C:\Users\User\Desktop\automation-myProjectName\index.mjs]
[C:\Users\User\Desktop\automation-myProjectName\index.coffee]
[C:\Users\User\Desktop\automation-myProjectName\index.ts]
[C:\Users\User\Desktop\automation-myProjectName\index.tsx]
 
    at Watching.handle [as handler] (C:\Users\User\AppData\Local\Cypress\Cache\7.3.0\Cypress\resources\app\packages\server\node_modules\@cypress\webpack-preprocessor\dist\index.js:172:23)
    at C:\Users\User\AppData\Local\Cypress\Cache\7.3.0\Cypress\resources\app\packages\server\node_modules\webpack\lib\Watching.js:99:9
    at AsyncSeriesHook.eval [as callAsync] (eval at create (C:\Users\User\AppData\Local\Cypress\Cache\7.3.0\Cypress\resources\app\packages\server\node_modules\tapable\lib\HookCodeFactory.js:33:10), <anonymous>:6:1)
    at AsyncSeriesHook.lazyCompileHook (C:\Users\User\AppData\Local\Cypress\Cache\7.3.0\Cypress\resources\app\packages\server\node_modules\tapable\lib\Hook.js:154:20)
    at Watching._done (C:\Users\User\AppData\Local\Cypress\Cache\7.3.0\Cypress\resources\app\packages\server\node_modules\webpack\lib\Watching.js:98:28)
    at C:\Users\User\AppData\Local\Cypress\Cache\7.3.0\Cypress\resources\app\packages\server\node_modules\webpack\lib\Watching.js:73:19
    at Compiler.emitRecords (C:\Users\User\AppData\Local\Cypress\Cache\7.3.0\Cypress\resources\app\packages\server\node_modules\webpack\lib\Compiler.js:499:39)
    at C:\Users\User\AppData\Local\Cypress\Cache\7.3.0\Cypress\resources\app\packages\server\node_modules\webpack\lib\Watching.js:54:20
    at C:\Users\User\AppData\Local\Cypress\Cache\7.3.0\Cypress\resources\app\packages\server\node_modules\webpack\lib\Compiler.js:485:14
    at AsyncSeriesHook.eval [as callAsync] (eval at create (C:\Users\User\AppData\Local\Cypress\Cache\7.3.0\Cypress\resources\app\packages\server\node_modules\tapable\lib\HookCodeFactory.js:33:10), <anonymous>:6:1)
    at AsyncSeriesHook.lazyCompileHook (C:\Users\User\AppData\Local\Cypress\Cache\7.3.0\Cypress\resources\app\packages\server\node_modules\tapable\lib\Hook.js:154:20)
    at C:\Users\User\AppData\Local\Cypress\Cache\7.3.0\Cypress\resources\app\packages\server\node_modules\webpack\lib\Compiler.js:482:27
    at C:\Users\User\AppData\Local\Cypress\Cache\7.3.0\Cypress\resources\app\packages\server\node_modules\neo-async\async.js:2818:7
    at done (C:\Users\User\AppData\Local\Cypress\Cache\7.3.0\Cypress\resources\app\packages\server\node_modules\neo-async\async.js:3522:9)
    at AsyncSeriesHook.eval [as callAsync] (eval at create (C:\Users\User\AppData\Local\Cypress\Cache\7.3.0\Cypress\resources\app\packages\server\node_modules\tapable\lib\HookCodeFactory.js:33:10), <anonymous>:6:1)
    at AsyncSeriesHook.lazyCompileHook (C:\Users\User\AppData\Local\Cypress\Cache\7.3.0\Cypress\resources\app\packages\server\node_modules\tapable\lib\Hook.js:154:20)
```

### Desired behavior
We do not want cypress to redirect on window.top.location.hash changes.

### Test code to reproduce
```js
var loc = window.location;
hash = hash.replace(hashRe, me.hashbang ? '#!' : '#');
 
if (replace) {
    loc.replace(hash);
} else {
    loc.hash = hash;
}
```

Our test does the following:

```js
cy.visit("https://localhost");
```

### Versions
We started having this issue while using cypress 5.3.0 and Chrome 89.
We have upgraded to cypress 7.3.0 and Chrome 91, but we still encounter the same, albeit the stack trace has some more data.
We also just upgraded to cypress 7.5.0 and Chrome 91, but the issue is still there.

Operating System: Windows 10

**Comment by cosminlupw** [comment]
+1 this happens to me as well it seems like cypress is having troubles sometimes wrapping location during the runner and it redirects outside of the app/test

**Comment by silviasolomon** [comment]
On my side is happening the same.

**Comment by jennifer-shehane** [comment]
Can someone provide a fully reproducible example of the issue? A repo would be ideal. The original comment does not contain all of the code necessary to reproduce. Thanks

**Comment by hiezust** [comment]
I cannot provide a fully working example, especially not a repo, but I can give further details of the situation we are encountering.

We make use of ExtJS framework and for keep tracking of history there is a built in component called [Ext.util.History](https://docs.sencha.com/extjs/6.5.2/modern/Ext.util.History.html#property-useTopWindow). On this component we make use of `useTopWindow` property since our application makes use of iframes.

Since we use this property, the `window.top.location.hash` is being used instead of `window.location.hash`, but the `window.top.location` is being overriden to be something like this:

```
https://localhost/__/#/tests/integration/module1/module/module-test.spec.ts
```

Usually, `window.top.location` would be
```
https://localhost/appEndpoint
```

From here the test goes downhill and we encounter the error from above.

**Comment by Aroxx** [comment]
I am also getting the same "Error: Webpack Compilation Error" posted above in an ExtJS application as soon as I click on anything within the application.

If I turn on both `modifyObstructiveCode` and `experimentalSourceRewriting` in my cypress.json file as per a suggestion in [#8507](https://github.com/cypress-io/cypress/issues/8507) I can get past the above error for one click but then I receive a different error (below) if I subsequently click anything else. If I only turn on one flag or the other then I still end up with the same Webpack error above on the first click within the app.

Second error with `modifyObstructiveCode` and `experimentalSourceRewriting` both set to true and clicking a second target:

`
TypeError
The following error originated from your application code, not from Cypress.
Uncaught TypeError: globalThis.top.Cypress.resolveWindowReference(...).on is not a function cypress_runner.js:141919
`

Unfortunately, I also cannot provide a reproducible sample because my app is not public. As hiezust mentioned, this appears to be an issue with the way ExtJS uses window.top.location.hash for routing not playing well with Cypress rewriting those attributes.

**Comment by batmaid-uciek** [comment]
I have the same issue every time on a redirect after payment and it's blocking our tests. Instead of `http://localhost/someRoute`, Cypress goes to `http://localhost/__/someRoute`as described above by others.

@jennifer-shehane is there any chance of this being resolved any time soon? Seeing this being reported back in September (https://github.com/cypress-io/cypress/issues/8507) does not inspire hope

**Comment by batmaid-uciek** [comment]
this issue still occurs in 7.7.0

**Comment by cosminlupw** [comment]
After some investigation this has to do with the way Ext.History api works when a change in location occurs (even when a # is added) and the fact that cypress does not wrap/rewrite/modify "window" objects in Ext Js framework files

The solution is to listen to the window onload event and set
```javascript
 Ext.History.useTopWindow = false;
```
This will prevent redirecting outside of the contentWindow.
My explanation is shallow in details, if you need more information you can set a debugger in a 
```javascript
window.unload = function(event) {
  debugger;
 }
```

and follow the Call Stack to where the redirect outside your app started. You will see that by default Ext.History.useTopWindow = true and it will use the "window.top" when chaning the location which at that point is actually the cypress runner window;

TLDR;
If you don't want to change your app codebase, because it really does not make sense to modify your codebase if under normal circumstances everything works normally, you can do it with cypress in one simple way that will avoid this issues when running specs automatically or when navigating manually through the Test Runner interface to your app.

in index.ts or index.js 
just add this

```javascript
Cypress.on("window:load", (win: any) => {
    if (win.Ext) {
        console.info("Global events index.ts: Ext is defined. Will set Ext.History.useTopWindow = false")
        win.Ext.History.useTopWindow = false;
    }
    else {
        console.info("Global events index.ts: Ext is not yet defined")
    }
});
```

Note:
I was using the below cypress config before this and at some point the redirect did not happen (don't know yet why it started to redirect outside), but now it works without them as well and speed of page loading has increased tremendously.
```json
{
...
 "modifyObstructiveCode": true,
  "experimentalSourceRewriting": true,
}

```

Have fun!

**Comment by JonasDominguini** [comment]
I was having a very similar problem. But I solved it in a simpler way, maybe because it's a more current version.
Systems that I'm automating is in EXT.JS

Putting in cypress.json
{
  "viewportWidth": 1440, 
  "viewportHeight": 900,
  "baseUrl": "your initial url"
"chromeWebSecurity": false
}

Version cypress 9.5.4
Windows: 10
Node-js version: 16.14.0

**Comment by RahulBhandari25** [comment]
This seems to be an issue in version 9.5.4 as well.
We are using property top.window.location.pathname and its overriden by cypress url which basically returns "/__/"

**Comment by joydeep100** [comment]
Yes we too are using ExtJs.

For us, adding the code in index.js file did not work. Was getting "window:load" event not allowed.

What worked instead was adding this piece of code in the spec file itself.

```
    Cypress.on('window:load', win => {
        if (win.Ext) 
            win.Ext.History.useTopWindow = false;
    });
```

**Comment by nagash77** [comment]
I am closing this issue until a reproducible example can be provided.
